File: rust/smooth-operator-core/src/executor.rs
//! Agent execution backend abstraction.
//!
//! An [`AgentExecutor`] is the seam that decides *where and how* an agent turn
//! runs, while [`Agent`](crate::agent::Agent) remains the unit of orchestration
//! (system prompt, tools, loop policy, compaction, checkpointing). Today there
//! is one implementation, [`InProcessExecutor`], which drives the turn directly
//! in the calling task — identical behavior to calling [`Agent::run`] /
//! [`Agent::run_with_channel`] yourself. It carries no dependencies and needs no
//! infrastructure: it is the zero-infra default and the OSS default.
//!
//! The point of the trait is to let an *optional* durable backend
//! (a Temporal-backed executor, in a separate crate, behind a feature flag and
//! off by default) plug in without the engine or its consumers knowing. Such a
//! backend models the turn as a durable workflow and runs the agent's
//! side-effects (model calls, tool invocations, retrieval, persistence) as
//! activities, giving crash-safe resume, durable human-in-the-loop via signals,
//! and durable timers — see ADR-030 (`SMOODEV-1972`). The in-process executor
//! runs those same side-effects inline; the trait boundary is what lets the two
//! be swapped at the edge.
//!
//! Keeping this as a thin dispatch seam (rather than reaching into the loop) is
//! deliberate: the in-process path stays a verbatim delegation to the existing,
//! battle-tested [`Agent::run`] surface, so introducing the abstraction changes
//! no behavior. The deeper orchestration-vs-activity split that a durable
//! backend needs lives behind this boundary and is built in the executor that
//! requires it, not forced onto the default path.

use tokio::sync::mpsc::UnboundedSender;

use async_trait::async_trait;

use crate::agent::{Agent, AgentEvent};
use crate::conversation::Conversation;

/// A backend that executes an [`Agent`] turn and returns the resulting
/// [`Conversation`].
///
/// Implementations decide *how* the turn is driven (inline, or on a durable
/// workflow engine); the [`Agent`] passed in owns *what* the turn does. Both
/// methods take `&Agent` so an executor never takes ownership of the agent — it
/// borrows it for the duration of the turn, exactly like [`Agent::run`].
#[async_trait]
pub trait AgentExecutor: Send + Sync {
    /// Run a single turn for `user_message`, returning the full conversation.
    ///
    /// Behaviorally equivalent to [`Agent::run`] for the in-process backend.
    ///
    /// # Errors
    /// Propagates any fatal error from the underlying turn (LLM call, tool
    /// execution, or — for durable backends — the workflow engine).
    async fn execute(&self, agent: &Agent, user_message: String) -> anyhow::Result<Conversation>;

    /// Run a single turn, emitting [`AgentEvent`]s over `events` as they occur
    /// (token deltas, tool start/complete, completion).
    ///
    /// Behaviorally equivalent to [`Agent::run_with_channel`] for the in-process
    /// backend.
    ///
    /// # Errors
    /// Propagates any fatal error from the underlying turn.
    async fn execute_streaming(&self, agent: &Agent, user_message: String, events: UnboundedSender<AgentEvent>) -> anyhow::Result<Conversation>;
}

/// The default, zero-infra executor: it drives the turn inline in the calling
/// task by delegating straight to [`Agent::run`] / [`Agent::run_with_channel`].
///
/// This is a verbatim pass-through — introducing it changes no behavior. It is
/// the executor used unless a consumer explicitly opts into a durable backend.
#[derive(Debug, Clone, Copy, Default)]
pub struct InProcessExecutor;

impl InProcessExecutor {
    /// Construct the in-process executor. (It is a unit type; `new` exists for
    /// symmetry with stateful executors.)
    #[must_use]
    pub fn new() -> Self {
        Self
    }
}

#[async_trait]
impl AgentExecutor for InProcessExecutor {
    async fn execute(&self, agent: &Agent, user_message: String) -> anyhow::Result<Conversation> {
        agent.run(user_message).await
    }

    async fn execute_streaming(&self, agent: &Agent, user_message: String, events: UnboundedSender<AgentEvent>) -> anyhow::Result<Conversation> {
        agent.run_with_channel(user_message, events).await
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;
    use crate::agent::AgentConfig;
    use crate::llm::{LlmConfig, StreamEvent};
    use crate::llm_provider::MockLlmClient;
    use crate::tool::ToolRegistry;

    fn test_config() -> AgentConfig {
        AgentConfig::new("test-agent", "You are a test agent", LlmConfig::openrouter("fake-key"))
    }

    /// The in-process executor drives the loop identically to `Agent::run`:
    /// a single text response with no tool calls ends the turn after one LLM
    /// call, with the assistant content the mock scripted.
    #[tokio::test]
    async fn in_process_executor_matches_agent_run() {
        let mock = MockLlmClient::new();
        mock.push_text("the answer is 42");
        let agent = Agent::new(test_config(), ToolRegistry::new()).with_llm_provider(Arc::new(mock.clone()));

        let convo = InProcessExecutor::new()
            .execute(&agent, "what is the answer?".to_string())
            .await
            .expect("execute completes");

        assert_eq!(convo.last_assistant_content(), Some("the answer is 42"));
        assert_eq!(mock.call_count(), 1);
        let calls = mock.calls();
        assert!(calls[0].messages.iter().any(|m| m.content.contains("what is the answer?")));
    }

    /// The streaming entry point surfaces events over the channel and returns
    /// the same final conversation.
    #[tokio::test]
    async fn in_process_executor_streaming_emits_events_and_returns_conversation() {
        // The streaming path drives `chat_stream`, which the mock services from
        // its `streams` queue — script it with deltas + a terminal Done.
        let mock = MockLlmClient::new();
        mock.push_stream(vec![
            StreamEvent::Delta { content: "streamed ".into() },
            StreamEvent::Delta { content: "reply".into() },
            StreamEvent::Done { finish_reason: "stop".into() },
        ]);
        let agent = Agent::new(test_config(), ToolRegistry::new()).with_llm_provider(Arc::new(mock.clone()));

        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
        let convo = InProcessExecutor::new()
            .execute_streaming(&agent, "stream please".to_string(), tx)
            .await
            .expect("streaming execute completes");

        assert_eq!(convo.last_assistant_content(), Some("streamed reply"));

        // At least one event was emitted (Started, ... Completed).
        let mut events = Vec::new();
        while let Ok(ev) = rx.try_recv() {
            events.push(ev);
        }
        assert!(!events.is_empty(), "expected streaming events to be emitted");
    }

    /// A dyn-dispatched executor works — the abstraction is object-safe, which
    /// is what lets a consumer hold `Arc<dyn AgentExecutor>` and swap backends.
    #[tokio::test]
    async fn executor_is_object_safe() {
        let mock = MockLlmClient::new();
        mock.push_text("dyn ok");
        let agent = Agent::new(test_config(), ToolRegistry::new()).with_llm_provider(Arc::new(mock.clone()));

        let executor: Arc<dyn AgentExecutor> = Arc::new(InProcessExecutor::new());
        let convo = executor.execute(&agent, "via dyn".to_string()).await.expect("dyn execute completes");

        assert_eq!(convo.last_assistant_content(), Some("dyn ok"));
    }
}

File: python/core/src/smooth_operator_core/executor.py
"""The durable-execution seam (ADR-030): *where* a turn runs, and the
deterministic orchestration it runs.

Two boundaries live here, and they are deliberately separate:

* :class:`AgentExecutor` — the coarse dispatch seam. It decides where and how a
  whole turn runs while :class:`~smooth_operator_core.agent.SmoothAgent` stays the
  unit of orchestration (instructions, tools, budget, compaction, checkpointing).
  The only implementation shipped here is :class:`InProcessExecutor`, a verbatim
  delegation to :meth:`SmoothAgent.run` / :meth:`SmoothAgent.run_stream` — so
  introducing the seam changes no behavior, carries no dependency and needs no
  infrastructure. Both methods take the agent as a parameter: an executor borrows
  the agent for the turn, it never owns it.

* :class:`AgentActivities` + :func:`drive_turn` — the fine-grained split. The
  *side-effecting* half (the model call, each tool invocation) is what a durable
  backend runs as retried, memoized activities; the *deterministic* half is the
  loop that decides "call the model → if it asked for tools, run them and loop →
  else stop." :func:`drive_turn` performs no I/O, reads no wall clock and draws no
  randomness, which is what makes it replay-safe: a durable workflow can run this
  very function, and the in-process path runs it too. One loop, two backends — that
  is what keeps a durable path from drifting away from the inline one.

Everything crossing the activity boundary is a plain, serializable value: messages
and tool specs are OpenAI-wire dicts, the model call returns the assistant message
as a dict, and a tool call is the wire-shape ``{"id", "type", "function"}`` dict the
model emitted. Nothing here holds a live client, socket or generator.

Scope, mirroring the Rust reference: :func:`drive_turn` reproduces the *core*
decision flow of :meth:`SmoothAgent.run` (snapshot context → model call → append
assistant message → stop-or-run-tools → append results → loop to
``max_iterations``). It deliberately omits the inline-runtime concerns a durable
backend models differently or that are follow-ups: stream event emission,
checkpointing (workflow history *is* the checkpoint), compaction, budget
enforcement, parallel tool dispatch, knowledge/memory injection. Those layer onto
this loop without changing its shape.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, AsyncIterator, Protocol

from .agent import AgentOptions, AgentRunResponse, SmoothAgent, StreamEvent
from .hooks import ToolResult
from .llm_provider import LlmProvider
from .thread import SmoothAgentThread


class AgentExecutor(Protocol):
    """A backend that executes one agent turn.

    Implementations decide *how* the turn is driven (inline, or on a durable
    workflow engine); the :class:`~smooth_operator_core.agent.SmoothAgent` passed in
    owns *what* the turn does. The signatures mirror the agent's own entry points
    exactly, so an executor is a drop-in for calling the agent directly.
    """

    async def execute(
        self,
        agent: SmoothAgent,
        message: str,
        history: list[dict[str, Any]] | None = None,
        thread: SmoothAgentThread | None = None,
    ) -> AgentRunResponse:
        """Run a single turn, returning the same result :meth:`SmoothAgent.run` would.

        Propagates any fatal error from the underlying turn (model call, tool
        dispatch, or — for a durable backend — the workflow engine).
        """
        ...

    def execute_streaming(
        self,
        agent: SmoothAgent,
        message: str,
        history: list[dict[str, Any]] | None = None,
        thread: SmoothAgentThread | None = None,
    ) -> AsyncIterator[StreamEvent]:
        """Run a single turn, yielding :data:`~smooth_operator_core.agent.StreamEvent`s
        as work happens (text deltas, tool call/result, one terminal ``DoneEvent``).

        Not a coroutine: like :meth:`SmoothAgent.run_stream` this *returns* the async
        iterator rather than awaiting a result.
        """
        ...


class InProcessExecutor:
    """The default, zero-infra executor: drives the turn inline in the calling task
    by delegating straight to :meth:`SmoothAgent.run` / :meth:`SmoothAgent.run_stream`.

    A verbatim pass-through — introducing it changes no behavior. It is the executor
    used unless a consumer explicitly opts into a durable backend.
    """

    async def execute(
        self,
        agent: SmoothAgent,
        message: str,
        history: list[dict[str, Any]] | None = None,
        thread: SmoothAgentThread | None = None,
    ) -> AgentRunResponse:
        return await agent.run(message, history, thread)

    def execute_streaming(
        self,
        agent: SmoothAgent,
        message: str,
        history: list[dict[str, Any]] | None = None,
        thread: SmoothAgentThread | None = None,
    ) -> AsyncIterator[StreamEvent]:
        return agent.run_stream(message, history, thread)


# TODO(ADR-030): the Temporal-backed AgentExecutor/AgentActivities belongs in a
# SEPARATE distribution behind an optional extra (mirroring the Rust reference's
# separate `smooth-operator-temporal` crate and its off-by-default `temporal` cargo
# feature), so this published package stays zero-infra and never pulls a
# `temporalio` dependency. It implements these same two Protocols — scheduling
# `model_call` / `tool_invoke` as activities — then calls `drive_turn` unchanged.
class AgentActivities(Protocol):
    """The side-effecting boundary of a turn: the model call and tool invocations.

    A durable backend runs each of these as an activity (retried upstream, memoized
    in history); the in-process backend runs them inline. Inputs and outputs are
    plain OpenAI-wire values — dicts and a :class:`~smooth_operator_core.hooks.ToolResult`
    — so an implementation may serialize them across an activity boundary unchanged.
    """

    async def model_call(self, messages: list[dict[str, Any]], tools: list[dict[str, Any]] | None) -> dict[str, Any]:
        """Invoke the model with the given context and visible tool specs, returning
        the assistant message in OpenAI wire shape (``role``/``content``, plus
        ``tool_calls`` when the model asked for any).

        Propagates a fatal model-call error (network, upstream rejection). A durable
        backend turns transient failures into activity retries before this surfaces.
        """
        ...

    async def tool_invoke(self, call: dict[str, Any]) -> ToolResult:
        """Execute a single wire-shape tool call, returning its result.

        Raises only on a fatal *dispatch* error. Tool **business** failures are
        reported on the returned result's
        :attr:`~smooth_operator_core.hooks.ToolResult.is_error` — never as an
        exception — mirroring how the agent's own dispatch feeds a failing tool back
        to the model instead of crashing the turn.
        """
        ...


@dataclass(frozen=True)
class TurnPolicy:
    """Loop policy for :func:`drive_turn`: the bound that keeps the orchestration
    terminating. Mirrors :attr:`~smooth_operator_core.agent.AgentOptions.max_iterations`
    — which defaults to 8 here, where the Rust reference's ``AgentConfig`` defaults
    to 50; each language keeps its own engine's default.
    """

    max_iterations: int = 8


async def drive_turn(
    activities: AgentActivities,
    messages: list[dict[str, Any]],
    tools: list[dict[str, Any]] | None,
    policy: TurnPolicy,
) -> None:
    """Drive one agent turn deterministically over ``activities``, mutating
    ``messages`` in place.

    ``messages`` must already be seeded with the system prompt, any prior turns and
    the current user message — exactly the state :meth:`SmoothAgent.run` holds when
    it enters its loop. On return it carries the appended assistant and tool
    messages.

    This performs no I/O, wall-clock read or RNG of its own — all of that is
    delegated to ``activities`` — so it is safe to run as durable workflow code
    (replay-deterministic) as well as inline. It propagates the first fatal error
    from either activity; an exhausted iteration budget is not an error, it simply
    returns what has accumulated so far.
    """
    for _iteration in range(1, policy.max_iterations + 1):
        context = list(messages)

        response = await activities.model_call(context, tools)

        content = response.get("content") or ""
        tool_calls = response.get("tool_calls") or []

        # The Rust push condition is content OR tool calls OR reasoning content;
        # the OpenAI wire shape this engine speaks carries no reasoning field.
        if content or tool_calls:
            assistant: dict[str, Any] = {"role": "assistant", "content": content}
            if tool_calls:
                assistant["tool_calls"] = tool_calls
            messages.append(assistant)

        if not tool_calls:
            return

        for call in tool_calls:
            result = await activities.tool_invoke(call)
            messages.append(
                {
                    "role": "tool",
                    "tool_call_id": call["id"],
                    "name": call["function"]["name"],
                    "content": result.content,
                }
            )


class InProcessActivities:
    """The default, zero-infra :class:`AgentActivities`: runs the model call through
    an :class:`~smooth_operator_core.llm_provider.LlmProvider` and tool calls through
    the configured tool set, inline in the calling task.

    It holds a :class:`~smooth_operator_core.agent.SmoothAgent` built from the same
    provider and options purely as the *dispatch surface* — that is where this engine
    keeps the equivalents of the Rust reference's ``LlmProvider`` and ``ToolRegistry``
    (retry/backoff, ``max_tokens`` clamping, metadata, then tool resolution with
    clearance, human gate and the tool hooks). Reusing it is what makes the activity
    path behave identically to the inline loop; the agent's own turn loop is not used.
    """

    def __init__(self, llm: LlmProvider, options: AgentOptions) -> None:
        self._agent = SmoothAgent(llm, options)

    async def model_call(self, messages: list[dict[str, Any]], tools: list[dict[str, Any]] | None) -> dict[str, Any]:
        response = await self._agent._call_model(messages, tools)
        choice = response.choices[0].message
        assistant: dict[str, Any] = {"role": "assistant", "content": choice.content or ""}
        if choice.tool_calls:
            assistant["tool_calls"] = [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {"name": tc.function.name, "arguments": tc.function.arguments},
                }
                for tc in choice.tool_calls
            ]
        return assistant

    async def tool_invoke(self, call: dict[str, Any]) -> ToolResult:
        function = call["function"]
        return await self._agent._dispatch_tool_result(function["name"], function.get("arguments") or "", None)
